### PR TITLE
unit test for Azure part

### PR DESCRIPTION
**Pull request by bangqipropel**
it is good for review, and coverage is 52%

### CODE REVIEW

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+			mockCtrl                        *gomock.Controller
+			mockAzureServiceHelper          *MockazureServicesHelper
+			mockazureNwIntfWrapper          *MockazureNwIntfWrapper
+			mockazureNsgWrapper             *MockazureNsgWrapper
+			mockazureAsgWrapper             *MockazureAsgWrapper
+			mockazureVirtualNetworksWrapper *MockazureVirtualNetworksWrapper
+			mockazureResourceGraph          *MockazureResourceGraphWrapper
+			mockazureService                *MockazureServiceClientCreateInterface
+		)
+		_ = account
```
Comment (shenmo3) [inline pkg/cloud-provider/cloudapi/azure/azure_security_test.go:75]: This shouldn't be needed, can you remove it? Also in azure_test.go there is a similar instance at line 76, can you remove that too? These should be the left over thing we forgot to remove in the K8s secret PR.

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+		testClientID                    = "ClientID"
+		testClientKey                   = "ClientKey"
+		testTenantID                    = "TenantID"
+		testRegion                      = "eastus"
+		testRG                          = "testRG"
+		nsgID                           = "nephe-ag-nsgID"
+		atAsgID                         = "nephe-at-atapplicationsgID"
+		atAsgName                       = "atapplicationsgID"
+		agAsgID                         = "nephe-ag-agapplicationsgID"
+		testPriority              int32 = 100
```
Comment (reachjainrahul) [inline pkg/cloud-provider/cloudapi/azure/azure_security_test.go:53]: do you need to set this?

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+		testCidrStr                     = "192.168.1.1/24"
+
+		testVnet01   = "testVnet01"
+		testVnetID01 = fmt.Sprintf("/subscriptions/%v/resourceGroups/%v/providers/Microsoft.Network/virtualNetworks/%v",
+			testSubID, testRG, testVnet01)
+		testVnet02   = "testVnet02"
+		testVnetID02 = fmt.Sprintf("/subscriptions/%v/resourceGroups/%v/providers/Microsoft.Network/virtualNetworks/%v",
+			testSubID, testRG, testVnet02)
+
+		testVnetPeer01   = "testVnetPeer01"
```
Comment (reachjainrahul) [inline pkg/cloud-provider/cloudapi/azure/azure_security_test.go:69]: lets avoid Peering test for now

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+			err = c.AddProviderAccount(fakeClient, account)
+			Expect(err).Should(BeNil())
+			selector.Spec.VMSelector = vmSelector
+			err = c.AddAccountResourceSelector(testAccountNamespacedName, selector)
+			Expect(err).Should(BeNil())
+
+			accCfg, _ := c.cloudCommon.GetCloudAccountByName(testAccountNamespacedName)
+			serviceConfig, _ := accCfg.GetServiceConfigByName(azureComputeServiceNameCompute)
+
+			vnetIDs := make(map[string]struct{})
```
Comment (reachjainrahul) [inline pkg/cloud-provider/cloudapi/azure/azure_security_test.go:229]: generally its not a good idea to combine everything in before test, unless its common for all. can we break this into smaller funcs/modules and explicitly call from the test. For example, define a function like setupAccount, addSelector etc.

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+							webAddressGroupIdentifier03,
+						},
+					},
+				}
+
+				err := c.UpdateSecurityGroupRules(webAddressGroupIdentifier03, ingressrules, egressrules)
+				Expect(err).Should(BeNil())
+			})
+
+			It("Should update Security rules for Peerings", func() {
```
Comment (reachjainrahul) [inline pkg/cloud-provider/cloudapi/azure/azure_security_test.go:344]: For now, remove peering test. We will add it when some coverage on integration tests too

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+						Name: *webAddressGroupIdentifier02,
+					},
+				}
+
+				err := c.UpdateSecurityGroupMembers(webAddressGroupIdentifier01, members, false)
+				Expect(err).Should(BeNil())
+			})
+		})
+
+		Context("UpdateSecurityRules", func() {
```
Comment (reachjainrahul) [inline pkg/cloud-provider/cloudapi/azure/azure_security_test.go:304]: could you add failure tests too

File: pkg/cloud-provider/cloudapi/azure/azure_test.go [hunk 7]
```diff
@@ -244,7 +261,160 @@ var _ = Describe("Azure", func() {
...
+			err = c.AddAccountResourceSelector(testAccountNamespacedName, selector)
+			Expect(err).Should(BeNil())
+
+			accCfg, _ := c.cloudCommon.GetCloudAccountByName(testAccountNamespacedName)
+			serviceConfig, _ := accCfg.GetServiceConfigByName(azureComputeServiceNameCompute)
+			filters := serviceConfig.(*computeServiceConfig).computeFilters[selector.Name]
+			Expect(filters).To(Equal(expectedQueryStrs))
+		})
+
+		It("Should match expected filter - single VM Name only match", func() {
```
Comment (reachjainrahul) [inline pkg/cloud-provider/cloudapi/azure/azure_test.go:349]: whats the difference between this test and test at line 265?

File: pkg/cloud-provider/cloudapi/azure/azure_test.go [hunk 8]
```diff
@@ -244,7 +261,160 @@ var _ = Describe("Azure", func() {
 			filters := serviceConfig.(*computeServiceConfig).computeFilters[selector.Name]
 			Expect(filters).To(Equal(expectedQueryStrs))
 		})
+
+		It("Should match expected filter - single VMID only match", func() {
```
Comment (reachjainrahul) [inline pkg/cloud-provider/cloudapi/azure/azure_test.go:265]: Try to add all combination. 
1) VNETID match
2) VMID match
3) VM Name Match
4) VNETID + VMID
5) VNETID + VMName
6) Update and rollback. Like start with VNETID, then add VM Match and then remove VM Match... and so on

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go (hunk 2 above)
Comment (bangqipropel) [reply pkg/cloud-provider/cloudapi/azure/azure_security_test.go:53]: I think nsgrule need this as int32

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go (hunk 3 above)
Comment (bangqipropel) [reply pkg/cloud-provider/cloudapi/azure/azure_security_test.go]: deleted

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go (hunk 5 above)
Comment (bangqipropel) [reply pkg/cloud-provider/cloudapi/azure/azure_security_test.go]: deleted

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go (hunk 6 above)
Comment (bangqipropel) [reply pkg/cloud-provider/cloudapi/azure/azure_security_test.go:282]: added

File: pkg/cloud-provider/cloudapi/azure/azure_security_test.go (hunk 4 above)
Comment (bangqipropel) [reply pkg/cloud-provider/cloudapi/azure/azure_security_test.go:223]: I think for asg and nsg part they can use for multiple times, it should be in the Return() part, and other create and addaccount part, it is always return an error in one line and then need a check error, I think this part also mostly the same steps in the aws tests

File: pkg/cloud-provider/cloudapi/azure/azure_test.go (hunk 8 above)
Comment (bangqipropel) [reply pkg/cloud-provider/cloudapi/azure/azure_test.go:265]: Here is the problem now, seems like our RemoveSelector is not completed, so we cannot do the rollback now

**Comment by bangqipropel** [comment]
After removing the Peer part, and added some other tests, now coverage is 50.7%

**Comment by bangqipropel** [comment]
addressed the comments and as discussed, added the peering part back, new coverage is 56.1%.

**Comment by bangqipropel** [comment]
@reachjainrahul Updated last week and sent the email, please review it, thanks

**Review by reachjainrahul — APPROVED** [review]
/LGTM

**Pull request merged**